src/typecheck.rs

```diff
@@ -109,337 +232,234 @@ impl PrimitiveLike for ValueEq {
-
-        // Do literals first
-        for (node, &id) in &self.nodes {
-            match node {
-                ENode::Literal(lit) => {
-                    let old = leaves.insert(id, Expr::Lit(lit.clone()));
-                    if let Some(Expr::Lit(old_lit)) = old {
-                        panic!("Duplicate literal: {:?} {:?}", old_lit, lit);
-                    }
-                }
-                _ => continue,
+    // figure 8 of the relational e-matching paper
+    fn flatten_expr(&mut self, expr: &Expr) -> (AtomTerm, Vec<Atom<Symbol>>) {
+        match expr {
+            Expr::Var(var) => {
+                let var = if self.egraph.global_bindings.get(var).is_some() {
+                    AtomTerm::Global(*var)
+                } else {
+                    AtomTerm::Var(*var)
+                };
+                (var, vec![])
             }
-        }
-        // Globally bound variables first
-        for (node, &id) in &self.nodes {
-            match node {
-                ENode::Var(var) => {
-                    if self.egraph.global_bindings.get(var).is_some() {
-                        match leaves.entry(id) {
-                            Entry::Occupied(existing) => {
-                                canon.insert(*var, existing.get().clone());
-                            }
-                            Entry::Vacant(v) => {
-                                v.insert(Expr::Var(*var));
-                            }
-                        }
-                    }
-                }
-                _ => continue,
+            Expr::Lit(lit) => (AtomTerm::Literal(lit.clone()), vec![]),
+            Expr::Call(f, args) => {
+                let children: Vec<_> = args.iter().map(|arg| self.flatten_expr(arg)).collect();
+                let id = self.unionfind.make_set();
+                let var = AtomTerm::Var(Symbol::from(format!("$v{}", id)));
+                let ats: Vec<_> = children
+                    .iter()
+                    .map(|c| c.0.clone())
+                    .chain(once(var.clone()))
+                    .collect();
+
+                let mut atoms: Vec<_> = children.into_iter().flat_map(|c| c.1).collect();
+                atoms.push(Atom {
+                    head: *f,
+                    args: ats,
+                });
+                (var, atoms)
             }
         }
+    }
 
-        // Now do variables
-        for (node, &id) in &self.nodes {
-            debug_assert_eq!(id, self.unionfind.find(id));
-            match node {
-                ENode::Var(var) => match leaves.entry(id) {
-                    Entry::Occupied(existing) => {
-                        canon.insert(*var, existing.get().clone());
-                    }
-                    Entry::Vacant(v) => {
-                        v.insert(Expr::Var(*var));
-                    }
-                },
-                _ => continue,
+    fn flatten_fact(&mut self, fact: &Fact) -> Query<Symbol> {
```

Comment:
Isn't this basically what desugaring already does? It seems like if you just use the existing `NormFact` you will be much closer to what you want.